File: backend/services/stigmer-server/pkg/domain/workflow/registry/model_registry_store_test.go
package registry

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"sync/atomic"
	"testing"
	"time"
)

// A minimal upstream registry document: one model the bundle does not have.
const upstreamRegistryDoc = `{
  "models": [
    {
      "id": "claude-nova-6",
      "apiModelId": "claude-nova-6-20270101",
      "provider": "anthropic",
      "harness": "native",
      "pricing": {"inputPricePerMillion": 1.0, "outputPricePerMillion": 5.0}
    }
  ]
}`

func newStoreFromEmbed(t *testing.T) *ModelRegistryStore {
	t.Helper()
	data, err := ReadEmbeddedModelRegistry()
	if err != nil {
		t.Fatalf("ReadEmbeddedModelRegistry failed: %v", err)
	}
	s := &ModelRegistryStore{}
	if err := s.applyDocument(data); err != nil {
		t.Fatalf("bundled registry failed the sanity gate: %v", err)
	}
	return s
}

func TestModelRegistryStore_BundleLoadsAndIndexes(t *testing.T) {
	s := newStoreFromEmbed(t)

	if !s.HasAnyModels() {
		t.Fatal("expected the bundled registry to index models")
	}
	if !s.HasHarness("native") || !s.HasHarness("cursor") {
		t.Fatal("expected both native and cursor harnesses in the bundle")
	}
	// The id -> apiModelId contract (stigmer/stigmer#240): both forms are valid.
	if !s.IsValidModel("native", "claude-sonnet-4.6") {
		t.Error("canonical id claude-sonnet-4.6 must be a valid native model")
	}
	if !s.IsValidModel("native", "claude-sonnet-4-6") {
		t.Error("api id claude-sonnet-4-6 must be a valid native model")
	}
	if s.IsValidModel("native", "not-a-model") {
		t.Error("unknown models must not validate")
	}
	// Suggestions surface canonical ids only.
	for _, id := range s.CanonicalModels("native") {
		if id == "claude-sonnet-4-6" {
			t.Error("CanonicalModels must not contain provider api ids")
		}
	}
}

func TestModelRegistryStore_PricingVariantCapabilityIndex(t *testing.T) {
	s := newStoreFromEmbed(t)

	// composer-2.5 carries pricingVariants.fast in the bundle — the
	// capability that makes SERVICE_TIER_FAST selectable (#357).
	if !s.HasPricingVariant("composer-2.5", "fast") {
		t.Error("composer-2.5 must price the fast variant in the bundled registry")
	}
	// A native-harness model with no variant block must fail closed.
	if s.HasPricingVariant("claude-sonnet-4.6", "fast") {
		t.Error("claude-sonnet-4.6 must not report a fast variant")
	}
	if s.HasPricingVariant("not-a-model", "fast") {
		t.Error("unknown models must not report variants")
	}
	if s.HasPricingVariant("composer-2.5", "turbo") {
		t.Error("unknown variant keys must not match")
	}

	withFast := s.CanonicalModelsWithVariant("fast")
	if len(withFast) == 0 {
		t.Fatal("expected at least one model pricing the fast variant")
	}
	found := false
	for i, id := range withFast {
		if id == "composer-2.5" {
			found = true
		}
		if i > 0 && withFast[i-1] > id {
			t.Errorf("CanonicalModelsWithVariant must be sorted; %q before %q", withFast[i-1], id)
		}
	}
	if !found {
		t.Error("CanonicalModelsWithVariant(\"fast\") must include composer-2.5")
	}
}

// The harness-scoped variant lookup backs the workflow validators: the fast
// variant must be priced FOR THE TASK'S HARNESS, so a variant priced only
// under cursor never validates a native task (the silent-no-op leak, #357).
func TestModelRegistryStore_PricingVariantHarnessScope(t *testing.T) {
	s := Store()

	if !s.HasPricingVariantForHarness("cursor", "composer-2.5", "fast") {
		t.Error("composer-2.5 must price the fast variant under the cursor harness")
	}
	// The bundle prices composer-2.5's fast variant under cursor only — the
	// native scope must fail closed even though the any-harness form passes.
	if s.HasPricingVariantForHarness("native", "composer-2.5", "fast") {
		t.Error("composer-2.5 must not report a fast variant under the native harness")
	}
	if s.HasPricingVariantForHarness("cursor", "not-a-model", "fast") {
		t.Error("unknown models must not report variants under any harness")
	}

	cursorFast := s.CanonicalModelsWithVariantForHarness("cursor", "fast")
	if len(cursorFast) == 0 {
		t.Fatal("expected cursor-harness models pricing the fast variant")
	}
	for i := 1; i < len(cursorFast); i++ {
		if cursorFast[i-1] > cursorFast[i] {
			t.Errorf("CanonicalModelsWithVariantForHarness must be sorted; %q before %q",
				cursorFast[i-1], cursorFast[i])
		}
	}
	if got := s.CanonicalModelsWithVariantForHarness("native", "fast"); len(got) != 0 {
		t.Errorf("expected no native-harness fast models in the bundle, got %v", got)
	}
}

// The harness-scoped capability lookup backs THINKING_MODE_ENABLED
// validation (stigmer/stigmer#772): the thinking capability must be
// declared FOR THE HARNESS THAT CAN HONOR IT. The bundle declares
// capabilities.thinking on cursor entries (claude-opus-4-6) and on native
// entries (claude-sonnet-4.6) — but v1 has no native thinking wire
// mapping, so validators query the cursor scope and a native-only
// declaration must not leak through.
func TestModelRegistryStore_CapabilityIndex(t *testing.T) {
	s := newStoreFromEmbed(t)

	if !s.HasCapabilityForHarness("cursor", "claude-opus-4-6", ThinkingCapabilityKey) {
		t.Error("claude-opus-4-6 must declare the thinking capability under the cursor harness")
	}
	// composer-2.5's capabilities block declares thinking=false — a false
	// flag must index nothing (only literal true declares the capability).
	if s.HasCapabilityForHarness("cursor", "composer-2.5", ThinkingCapabilityKey) {
		t.Error("composer-2.5 declares thinking=false and must not index the capability")
	}
	// claude-sonnet-4.6 declares thinking on its NATIVE entry only.
	if !s.HasCapabilityForHarness("native", "claude-sonnet-4.6", ThinkingCapabilityKey) {
		t.Error("claude-sonnet-4.6 must declare the thinking capability under the native harness")
	}
	if s.HasCapabilityForHarness("cursor", "claude-sonnet-4.6", ThinkingCapabilityKey) {
		t.Error("claude-sonnet-4.6 has no cursor entry — the cursor scope must fail closed")
	}
	// The id -> apiModelId contract (stigmer/stigmer#240) extends to the
	// capability index: both reference forms resolve.
	if !s.HasCapabilityForHarness("native", "claude-sonnet-4-6", ThinkingCapabilityKey) {
		t.Error("api id claude-sonnet-4-6 must resolve the native thinking capability")
	}
	if s.HasCapabilityForHarness("cursor", "not-a-model", ThinkingCapabilityKey) {
		t.Error("unknown models must not report capabilities")
	}
	if s.HasCapabilityForHarness("cursor", "claude-opus-4-6", "levitation") {
		t.Error("unknown capability keys must not match")
	}

	cursorThinking := s.CanonicalModelsWithCapabilityForHarness("cursor", ThinkingCapabilityKey)
	if len(cursorThinking) == 0 {
		t.Fatal("expected cursor-harness models declaring the thinking capability")
	}
	found := false
	for i, id := range cursorThinking {
		if id == "claude-opus-4-6" {
			found = true
		}
		if i > 0 && cursorThinking[i-1] > id {
			t.Errorf("CanonicalModelsWithCapabilityForHarness must be sorted; %q before %q",
				cursorThinking[i-1], id)
		}
	}
	if !found {
		t.Error("cursor thinking capability list must include claude-opus-4-6")
	}
}

func TestModelRegistryStore_RefreshAppliesUpstreamToDocumentAndValidation(t *testing.T) {
	upstream := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != publicModelRegistryPath {
			t.Errorf("expected fetch of %s, got %s", publicModelRegistryPath, r.URL.Path)
		}
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(upstreamRegistryDoc))
	}))
	defer upstream.Close()

	s := newStoreFromEmbed(t)
	if s.IsValidModel("native", "claude-nova-6") {
		t.Fatal("fixture assumption: the bundle must not know claude-nova-6")
	}

	s.refreshOnce(context.Background(), upstream.URL+publicModelRegistryPath)

	// The served document IS the upstream document...
	var doc modelRegistryData
	if err := json.Unmarshal(s.Document(), &doc); err != nil {
		t.Fatalf("refreshed document is not valid JSON: %v", err)
	}
	if len(doc.Models) != 1 || doc.Models[0].ID != "claude-nova-6" {
		t.Fatalf("expected the upstream document to be served, got %d models", len(doc.Models))
	}
	// ...and validation moved with it — the endpoint and the valid-model
	// sets can never drift (DD-004).
	if !s.IsValidModel("native", "claude-nova-6") {
		t.Error("a refreshed model must validate")
	}
	if !s.IsValidModel("native", "claude-nova-6-20270101") {
		t.Error("a refreshed model's api id must validate")
	}
	if s.IsValidModel("native", "claude-sonnet-4.6") {
		t.Error("the store must swap, not merge — bundle models are gone after a refresh")
	}
}

func TestModelRegistryStore_UnreachableUpstreamKeepsCurrentRegistry(t *testing.T) {
	s := newStoreFromEmbed(t)
	before := s.Document()

	// A closed server: connection refused immediately.
	dead := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	deadURL := dead.URL
	dead.Close()

	s.refreshOnce(context.Background(), deadURL+publicModelRegistryPath)

	if string(s.Document()) != string(before) {
		t.Error("an unreachable upstream must not change the served registry")
	}
	if !s.IsValidModel("native", "claude-sonnet-4.6") {
		t.Error("validation must keep working from the previous registry")
	}
}

func TestModelRegistryStore_MalformedUpstreamIsRejected(t *testing.T) {
	cases := map[string]string{
		"not json":     `registry? what registry`,
		"empty models": `{"models": []}`,
		"no ids":       `{"models": [{"$comment": "only dividers"}]}`,
	}
	for name, body := range cases {
		t.Run(name, func(t *testing.T) {
			upstream := httptest.NewServer(http.HandlerFunc(
				func(w http.ResponseWriter, r *http.Request) {
					w.Write([]byte(body))
				}))
			defer upstream.Close()

			s := newStoreFromEmbed(t)
			before := s.Document()
			s.refreshOnce(context.Background(), upstream.URL+publicModelRegistryPath)

			if string(s.Document()) != string(before) {
				t.Error("a response failing the sanity gate must never replace the registry")
			}
		})
	}
}

func TestModelRegistryStore_ErrorStatusIsRejected(t *testing.T) {
	upstream := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "nope", http.StatusServiceUnavailable)
	}))
	defer upstream.Close()

	s := newStoreFromEmbed(t)
	before := s.Document()
	s.refreshOnce(context.Background(), upstream.URL+publicModelRegistryPath)

	if string(s.Document()) != string(before) {
		t.Error("a non-200 upstream must not change the served registry")
	}
}

func TestModelRegistryStore_RefreshOffMakesNoRequests(t *testing.T) {
	var requests atomic.Int32
	upstream := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requests.Add(1)
		w.Write([]byte(upstreamRegistryDoc))
	}))
	defer upstream.Close()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	s := newStoreFromEmbed(t)
	s.startRefresh(ctx, upstream.URL, "off")

	// The refresh goroutine, had it started, fetches immediately — give it
	// ample time to prove it did not.
	time.Sleep(150 * time.Millisecond)
	if got := requests.Load(); got != 0 {
		t.Fatalf("refresh=off must make zero outbound calls, saw %d", got)
	}
}

func TestModelRegistryStore_StartRefreshFetchesImmediately(t *testing.T) {
	applied := make(chan struct{})
	upstream := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(upstreamRegistryDoc))
	}))
	defer upstream.Close()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	s := newStoreFromEmbed(t)
	s.startRefresh(ctx, upstream.URL, "")

	// Poll (no fixed sleep): the first fetch is immediate, not TTL-delayed.
	go func() {
		deadline := time.Now().Add(5 * time.Second)
		for time.Now().Before(deadline) {
			if s.IsValidModel("native", "claude-nova-6") {
				close(applied)
				return
			}
			time.Sleep(10 * time.Millisecond)
		}
	}()

	select {
	case <-applied:
	case <-time.After(6 * time.Second):
		t.Fatal("the initial refresh did not apply within the deadline")
	}
}

File: backend/services/stigmer-server/src/domain/workflow/registry/__tests__/model-registry-store.test.ts
/**
 * Model-registry store unit tests, pinning the Go semantics
 * (model_registry_store.go): the sanity gate (parse + ≥1 INDEXED model —
 * id and harness both present, the applyDocument rule), keep-current-on-
 * failure, warn-then-debug failure logging with reset on success, the size
 * cap, the fail-loud bundled-snapshot guard, and the catalog indexes the
 * workflow validators query (harness/variant/capability, canonical
 * suggestion pools, api-id acceptance).
 */
import { describe, expect, it } from "vitest";

import { createLogger } from "../../../../boot/logger.js";
import {
  FAST_VARIANT_KEY,
  MODEL_REGISTRY_MAX_BYTES,
  ModelRegistryStore,
  THINKING_CAPABILITY_KEY,
} from "../model-registry-store.js";

const VALID_BUNDLE = JSON.stringify({
  models: [{ id: "anthropic/claude", harness: "native" }],
});
const UPGRADED = JSON.stringify({
  models: [
    { id: "anthropic/claude", harness: "native" },
    { id: "openai/gpt", harness: "native" },
  ],
});

function capturingLogger() {
  const lines: Array<{ level: string; message: string }> = [];
  return {
    lines,
    logger: createLogger({
      level: "debug",
      pretty: false,
      write: (line) =>
        lines.push(JSON.parse(line) as { level: string; message: string }),
    }),
  };
}

function store(
  fetchImpl: typeof fetch,
  logger = capturingLogger().logger,
): ModelRegistryStore {
  return new ModelRegistryStore({
    bundledDocument: VALID_BUNDLE,
    upstreamOrigin: "http://upstream.test",
    refreshEnabled: true,
    logger,
    fetchImpl,
  });
}

describe("ModelRegistryStore", () => {
  it("refuses to construct on an invalid bundled snapshot (build defect, fail loud)", () => {
    expect(
      () =>
        new ModelRegistryStore({
          bundledDocument: JSON.stringify({ models: [] }),
          upstreamOrigin: "http://upstream.test",
          refreshEnabled: true,
          logger: capturingLogger().logger,
        }),
    ).toThrow(/no models/);
  });

  it("applies an upstream document that passes the sanity gate", async () => {
    const subject = store(async () => new Response(UPGRADED, { status: 200 }));

    await subject.refreshOnce();

    expect(subject.document()).toBe(UPGRADED);
  });

  it.each([
    ["non-JSON body", "not json at all"],
    ["empty models array", JSON.stringify({ models: [] })],
    ["missing models key", JSON.stringify({ descriptors: [] })],
    // $comment section dividers carry no id/harness and index nothing — a
    // divider-only document is as unusable as an empty one (Go applyDocument
    // gates on the INDEX, not the raw array length).
    [
      "divider-only entries",
      JSON.stringify({ models: [{ $comment: "── Anthropic ──" }] }),
    ],
  ])(
    "keeps the current document when the upstream serves %s",
    async (_case, body) => {
      const subject = store(async () => new Response(body, { status: 200 }));

      await subject.refreshOnce();

      expect(subject.document()).toBe(VALID_BUNDLE);
    },
  );

  it("keeps the current document on upstream errors and oversized bodies", async () => {
    const failing = store(async () => new Response("nope", { status: 503 }));
    await failing.refreshOnce();
    expect(failing.document()).toBe(VALID_BUNDLE);

    const oversized = store(
      async () =>
        new Response(
          `{"models":[{"pad":"${"x".repeat(MODEL_REGISTRY_MAX_BYTES)}"}]}`,
          {
            status: 200,
          },
        ),
    );
    await oversized.refreshOnce();
    expect(oversized.document()).toBe(VALID_BUNDLE);
  });

  it("logs the first consecutive failure at warn, repeats at debug, resets on success", async () => {
    const { lines, logger } = capturingLogger();
    let mode: "fail" | "ok" = "fail";
    const subject = store(async () => {
      if (mode === "fail") {
        return new Response("down", { status: 500 });
      }
      return new Response(UPGRADED, { status: 200 });
    }, logger);

    await subject.refreshOnce();
    await subject.refreshOnce();
    expect(lines.map((line) => line.level)).toEqual(["warn", "debug"]);

    mode = "ok";
    await subject.refreshOnce();
    expect(subject.document()).toBe(UPGRADED);

    // The flag reset: the NEXT failure is "first" again → warn.
    mode = "fail";
    await subject.refreshOnce();
    expect(lines.map((line) => line.level)).toEqual(["warn", "debug", "warn"]);
  });
});

/**
 * The catalog indexes (Go applyDocument :302-412 + the query methods) — the
 * surface the workflow validators consume. Fixture shape mirrors real
 * registry entries: canonical id + provider api id, harness sections,
 * pricingVariants key set, tri-state capabilities.
 */
describe("ModelRegistryStore catalog indexes", () => {
  const CATALOG = JSON.stringify({
    models: [
      { $comment: "── section divider (indexes nothing) ──" },
      {
        id: "anthropic/claude-4",
        apiModelId: "claude-4-20260101",
        harness: "native",
        pricingVariants: { fast: { input: 1 } },
      },
      {
        id: "anthropic/claude-4",
        apiModelId: "claude-4-cursor",
        harness: "cursor",
        capabilities: { thinking: true },
      },
      {
        id: "openai/gpt-6",
        harness: "cursor",
        // false and non-boolean shapes index nothing (tri-state rule).
        capabilities: { thinking: false, vision: "yes" },
      },
      { id: "zeta/z-1", harness: "native" },
    ],
  });

  function catalogStore(): ModelRegistryStore {
    return new ModelRegistryStore({
      bundledDocument: CATALOG,
      upstreamOrigin: "http://upstream.test",
      refreshEnabled: false,
      logger: capturingLogger().logger,
    });
  }

  it("validates canonical ids AND provider api ids per harness (oss#240)", () => {
    const s = catalogStore();
    expect(s.isValidModel("native", "anthropic/claude-4")).toBe(true);
    expect(s.isValidModel("native", "claude-4-20260101")).toBe(true);
    expect(s.isValidModel("native", "claude-4-cursor")).toBe(false);
    expect(s.isValidModel("cursor", "claude-4-cursor")).toBe(true);
    expect(s.isValidModel("native", "openai/gpt-6")).toBe(false);
    expect(s.hasHarness("native")).toBe(true);
    expect(s.hasHarness("cli")).toBe(false);
    expect(s.hasAnyModels()).toBe(true);
  });

  it("answers the any-harness existence check and its merged suggestion pool", () => {
    const s = catalogStore();
    expect(s.isValidModelOnAnyHarness("openai/gpt-6")).toBe(true);
    expect(s.isValidModelOnAnyHarness("nope/never")).toBe(false);
    // Sorted, deduplicated canonical ids (claude-4 appears in two sections).
    expect(s.canonicalModelsAcrossHarnesses()).toEqual([
      "anthropic/claude-4",
      "openai/gpt-6",
      "zeta/z-1",
    ]);
    expect(s.canonicalModels("native")).toEqual([
      "anthropic/claude-4",
      "zeta/z-1",
    ]);
  });

  it("indexes pricing variants by key presence, harness-scoped and unioned", () => {
    const s = catalogStore();
    expect(
      s.hasPricingVariantForHarness("native", "anthropic/claude-4", FAST_VARIANT_KEY),
    ).toBe(true);
    // Priced under native only — the cursor section must not validate it
    // (the #357 silent-no-op class).
    expect(
      s.hasPricingVariantForHarness("cursor", "anthropic/claude-4", FAST_VARIANT_KEY),
    ).toBe(false);
    expect(s.hasPricingVariant("anthropic/claude-4", FAST_VARIANT_KEY)).toBe(true);
    expect(s.hasPricingVariant("zeta/z-1", FAST_VARIANT_KEY)).toBe(false);
    expect(s.canonicalModelsWithVariant(FAST_VARIANT_KEY)).toEqual([
      "anthropic/claude-4",
    ]);
    expect(
      s.canonicalModelsWithVariantForHarness("native", FAST_VARIANT_KEY),
    ).toEqual(["anthropic/claude-4"]);
    expect(
      s.canonicalModelsWithVariantForHarness("cursor", FAST_VARIANT_KEY),
    ).toEqual([]);
  });

  it("indexes only literal-true capabilities, harness-scoped (oss#772)", () => {
    const s = catalogStore();
    expect(
      s.hasCapabilityForHarness("cursor", "anthropic/claude-4", THINKING_CAPABILITY_KEY),
    ).toBe(true);
    expect(
      s.hasCapabilityForHarness("cursor", "claude-4-cursor", THINKING_CAPABILITY_KEY),
    ).toBe(true);
    // Declared under cursor only.
    expect(
      s.hasCapabilityForHarness("native", "anthropic/claude-4", THINKING_CAPABILITY_KEY),
    ).toBe(false);
    // false / non-boolean shapes never index (tri-state).
    expect(
      s.hasCapabilityForHarness("cursor", "openai/gpt-6", THINKING_CAPABILITY_KEY),
    ).toBe(false);
    expect(s.hasCapabilityForHarness("cursor", "openai/gpt-6", "vision")).toBe(false);
    expect(
      s.canonicalModelsWithCapabilityForHarness("cursor", THINKING_CAPABILITY_KEY),
    ).toEqual(["anthropic/claude-4"]);
  });

  it("swaps document and indexes together on refresh", async () => {
    const upgraded = JSON.stringify({
      models: [{ id: "new/model", harness: "native" }],
    });
    const s = new ModelRegistryStore({
      bundledDocument: CATALOG,
      upstreamOrigin: "http://upstream.test",
      refreshEnabled: true,
      logger: capturingLogger().logger,
      fetchImpl: async () => new Response(upgraded, { status: 200 }),
    });

    await s.refreshOnce();

    expect(s.document()).toBe(upgraded);
    expect(s.isValidModel("native", "new/model")).toBe(true);
    expect(s.isValidModel("native", "anthropic/claude-4")).toBe(false);
  });
});
